R/CITEseq.R

```diff
@@ -1,18 +1,145 @@
+                    m1 <- Matrix::Matrix(unlist(explist[assIdx]), 
+                                nrow=dimslist[assIdx][[1]][1],
+                                ncol=(dimslist[assIdx][[1]][2]+dimslist[assIdx][[2]][2]), 
+                                sparse=TRUE)
+                } else {
+                    m1 <- Matrix::Matrix(explist[[assIdx]])
+                }
+            },
+            "scRNA"={
+                if(length(explist[assIdx]) == 2)
+                {
+                    ## we can have at last 2 matrices
+                    m1 <- cbind(explist[[assIdx[1]]], explist[[assIdx[2]]])
+                    
+                } else {
+                    m1 <- explist[[assIdx]]
+                }
+            },
+            {stop("Unrecognized assayId: ", assayId)}
+            )
+        if(length(explist[assIdx]) == 2)
+        {
+            colnames(m1) <- c(paste0(rep(gsub("scADT|scHTO|scRNA","", 
+                                            names(explist)[assIdx[1]]),
+                                        dimslist[assIdx][[1]][2]), 
+                                    colnames(explist[[assIdx[1]]])),
+                              paste0(rep(gsub("scADT|scHTO|scRNA","", 
+                                            names(explist)[assIdx[2]]),
+                                        dimslist[assIdx][[2]][2]), 
+                                    colnames(explist[[assIdx[2]]])))
+        } else {
+            colnames(m1) <- paste0(rep(gsub("scADT|scHTO|scRNA","", 
+                                            names(explist)[assIdx[1]]),
+                                        dimslist[assIdx][[1]][2]), 
+                                    colnames(explist[[assIdx[1]]]))
+        }
+        return(m1)
+    }
+    
+    .buildMap <- function(mat1, assayId)
+    {
+        
+        map <- DataFrame(assay=assayId,
+                        #primary=gsub("_\\w+", "", colnames(mat1)), 
+                        primary=colnames(mat1),
+                        colname=colnames(mat1), 
+                        condition=gsub("_\\w+", "", colnames(mat1)))
+        return(map)
+    }
+    
+    # .buildColDat <- function(ll)
+    # {
+    #     if(all(names(ll)[grep("CTCL|CTRL", names(ll))] %in% names(ll)))
+    #     {
+    #         cd <- data.frame(row.names=c("CTCL", "CTRL"), 
+    #                    condition=c("Cutaneous T-cell Limphoma", "Control"))
+    #     } else if(!isEmpty(grep("CTCL", names(ll)))) {
+    #         cd <- data.frame(row.names=c("CTCL"), 
+    #                           condition=c("Cutaneous T-cell Limphoma"))
+    #     } else if(!isEmpty(grep("CTRL", names(ll)))) {
+    #         cd <- data.frame(row.names=c("CTRL"), 
+    #                           condition=c("Control"))
+    #     }
+    #     return(cd)
+    # }
+    
+    ll <- ess_list$experiments
+    ll <- lapply(ll, function(x) 
+    {
+        x <- x[order(rownames(x)),]
+    })
+    dims <- lapply(ll, dim)
+    expslist <- c()
```

Comment:
avoid growing vectors with `c(list, list)`, use the 'allocate and fill' method or even `mapply` on the selected assays, i.e., `grep("scADT", names(ll))`

Reply:
where do I find the "allocate and fill" methods for lists in R?